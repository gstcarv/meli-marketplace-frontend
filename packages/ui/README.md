# UI

Esse pacote contÃ©m os componentes base para a construÃ§Ãµes dos microfrontends.

## ğŸ“¦ InstalaÃ§Ã£o e ConfiguraÃ§Ã£o

### PrÃ©-requisitos

-   Node.js (versÃ£o 18.18.0+)
-   yarn

### Como rodar o storybook?

1. Instale as dependÃªncias:

    ```bash
    yarn install
    ```

2. Execute o storybook:

    ```bash
    yarn storybook
    ```

## ğŸ§ª Testes

Para rodar os testes, execute:

```bash
yarn test
```
