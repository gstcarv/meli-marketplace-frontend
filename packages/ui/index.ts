import './src/declarations/styled-components.d.ts';

export * from './src/components';

export * from './src/theme';
