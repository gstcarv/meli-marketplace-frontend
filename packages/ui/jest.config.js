module.exports = {
    testEnvironment: 'jsdom',
    transform: {
        '^.+\\.(j|t)sx?$': 'babel-jest'
    },
    moduleNameMapper: {
        '\\.(css)$': 'identity-obj-proxy',
        '\\.(png|jpg|jpeg|gif)$': 'identity-obj-proxy'
    },
    setupFilesAfterEnv: ['@testing-library/jest-dom', '<rootDir>/src/setupTests.ts']
};
