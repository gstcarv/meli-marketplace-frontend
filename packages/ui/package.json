{
    "name": "@meli/ui",
    "version": "0.0.0",
    "private": true,
    "main": "./index.ts",
    "scripts": {
        "lint": "eslint . --max-warnings 0",
        "test": "jest"
    },
    "devDependencies": {
        "@babel/preset-env": "^7.25.8",
        "@babel/preset-react": "^7.25.7",
        "@babel/preset-typescript": "^7.25.7",
        "@meli/design-tokens": "*",
        "@meli/eslint-config": "*",
        "@meli/icons": "*",
        "@meli/typescript-config": "*",
        "@testing-library/jest-dom": "^5.17.0",
        "@testing-library/react": "^12.1.2",
        "@turbo/gen": "^1.12.4",
        "@types/eslint": "^8.56.5",
        "@types/jest": "^27.0.1",
        "@types/node": "^20.11.24",
        "@types/react": "^18.2.61",
        "@types/react-dom": "^18.2.19",
        "babel-jest": "^29.7.0",
        "eslint": "^8.57.0",
        "jest": "^29.7.0",
        "jest-cli": "^29.7.0",
        "jest-environment-jsdom": "^29.7.0",
        "react-test-renderer": "^18.3.1",
        "typescript": "^5.3.3"
    },
    "peerDependencies": {
        "react": "^17.0.2",
        "styled-components": "^6.1.13"
    }
}
