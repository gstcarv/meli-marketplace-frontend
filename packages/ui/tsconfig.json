{
    "extends": "@meli/typescript-config/react-library.json",
    "compilerOptions": {
        "outDir": "dist",
        "module": "ES6",
        "moduleResolution": "Bundler"
    },
    "include": ["src", "./index.ts", "../design-tokens/src/lib"],
    "exclude": ["node_modules", "dist"]
}
