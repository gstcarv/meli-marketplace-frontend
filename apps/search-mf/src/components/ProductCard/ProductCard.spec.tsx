import { ProductItem } from '@marketplace/bff';
import { ThemeProvider } from '@meli/ui';
import { fireEvent, render, screen, waitFor } from '@testing-library/react';
import { MemoryRouter, useLocation } from 'react-router-dom';
import { ProductCard } from '.';

const productStub: Partial<ProductItem> = {
    id: 'mock-id',
    condition: 'new',
    title: 'Mocked product title',
    price: {
        amount: 4445,
        currency: 'ARS'
    },
    free_shipping: true,
    picture: 'http://mock/picture.png'
};

describe('<ProductCard />', () => {
    it('should render card correcly', () => {
        const { container } = render(
            <ThemeProvider>
                <MemoryRouter>
                    <ProductCard product={productStub} />
                </MemoryRouter>
            </ThemeProvider>
        );

        expect(screen.getByText('$ 4.445')).toBeInTheDocument();
        expect(screen.getByText('Mocked product title')).toBeInTheDocument();
        expect(screen.getByText('Produto novo')).toBeInTheDocument();

        expect(screen.getByRole('generic', { name: 'Frete grátis' })).toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it('should hide shipping icon if is not free shipping', () => {
        const { container } = render(
            <ThemeProvider>
                <MemoryRouter>
                    <ProductCard product={{ ...productStub, free_shipping: false }} />
                </MemoryRouter>
            </ThemeProvider>
        );

        expect(screen.queryByRole('generic', { name: 'Frete grátis' })).not.toBeInTheDocument();

        expect(container).toMatchSnapshot();
    });

    it('should navigate to product page when click card', async () => {
        const LocationPath = () => {
            const location = useLocation();

            return <div data-testid='pathname'>{location.pathname}</div>;
        };

        render(
            <ThemeProvider>
                <MemoryRouter>
                    <ProductCard product={{ ...productStub, free_shipping: false }} />

                    <LocationPath />
                </MemoryRouter>
            </ThemeProvider>
        );

        const card = screen.getByRole('link', { name: /product title/ });

        fireEvent.click(card);

        await waitFor(() => {
            expect(screen.getByTestId('pathname')).toHaveTextContent('/mock-id');
        });
    });
});
