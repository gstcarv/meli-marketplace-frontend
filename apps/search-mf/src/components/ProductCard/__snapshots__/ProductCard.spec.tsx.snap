// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProductCard /> should hide shipping icon if is not free shipping 1`] = `
.c0 {
  display: flex;
  flex-direction: row;
  padding: 2.75rem 0;
}

.c0:hover .ProductCard__title {
  color: #4A7FEC;
}

.c1 {
  width: 160px;
  height: 160px;
  border-radius: 0.25rem;
  display: flex;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  flex-shrink: 0;
}

.c1:empty {
  background: #EEEEEE;
}

.c1 img {
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.c2 {
  padding: 1rem;
  width: 50%;
}

.c3 {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.c4 {
  font-size: 1.375rem;
  font-weight: 400;
}

.c5 {
  font-size: 1.125rem;
  font-weight: 300;
  margin-top: 1rem;
}

.c6 {
  flex: 1;
}

.c7 {
  font-weight: 300;
  font-size: 0.875rem;
  color: #727272;
  display: block;
  text-align: right;
  padding-right: 20%;
}

@media screen and (max-width: 576px) {
  .c0 {
    flex-direction: column;
  }
}

@media screen and (max-width: 576px) {
  .c1 {
    width: 100%;
    height: 250px;
  }
}

@media screen and (max-width: 576px) {
  .c2 {
    width: 100%;
  }
}

@media screen and (max-width: 576px) {
  .c7 {
    text-align: left;
    padding: 0 1rem;
  }
}

<div>
  <a
    class="c0"
    href="/mock-id"
  >
    <div
      class="c1"
    >
      <img
        loading="lazy"
        src="http://mock/picture.png"
      />
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <p
          class="c4"
        >
          $ 4.445
        </p>
      </div>
      <h2
        class="c5 ProductCard__title"
      >
        Mocked product title
      </h2>
    </div>
    <div
      class="c6"
    >
      <span
        class="c7"
      >
        Produto novo
      </span>
    </div>
  </a>
</div>
`;

exports[`<ProductCard /> should render card correcly 1`] = `
.c5 {
  width: 22px;
  height: 22px;
  border-radius: 9999px;
  background: #00a650;
  display: flex;
  justify-content: center;
  align-items: center;
}

.c0 {
  display: flex;
  flex-direction: row;
  padding: 2.75rem 0;
}

.c0:hover .ProductCard__title {
  color: #4A7FEC;
}

.c1 {
  width: 160px;
  height: 160px;
  border-radius: 0.25rem;
  display: flex;
  justify-content: center;
  align-items: center;
  overflow: hidden;
  flex-shrink: 0;
}

.c1:empty {
  background: #EEEEEE;
}

.c1 img {
  width: 100%;
  height: 100%;
  object-fit: contain;
}

.c2 {
  padding: 1rem;
  width: 50%;
}

.c3 {
  display: flex;
  align-items: center;
  gap: 0.5rem;
}

.c4 {
  font-size: 1.375rem;
  font-weight: 400;
}

.c6 {
  font-size: 1.125rem;
  font-weight: 300;
  margin-top: 1rem;
}

.c7 {
  flex: 1;
}

.c8 {
  font-weight: 300;
  font-size: 0.875rem;
  color: #727272;
  display: block;
  text-align: right;
  padding-right: 20%;
}

@media screen and (max-width: 576px) {
  .c0 {
    flex-direction: column;
  }
}

@media screen and (max-width: 576px) {
  .c1 {
    width: 100%;
    height: 250px;
  }
}

@media screen and (max-width: 576px) {
  .c2 {
    width: 100%;
  }
}

@media screen and (max-width: 576px) {
  .c8 {
    text-align: left;
    padding: 0 1rem;
  }
}

<div>
  <a
    class="c0"
    href="/mock-id"
  >
    <div
      class="c1"
    >
      <img
        loading="lazy"
        src="http://mock/picture.png"
      />
    </div>
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        <p
          class="c4"
        >
          $ 4.445
        </p>
        <div
          aria-label="Frete grÃ¡tis"
          class="c5"
          color="success-500"
        >
          <svg
            class="icon-truck"
            color="#ffffff"
            fill="none"
            height="11"
            size="11"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="11"
            xmlns="http://www.w3.org/2000/svg"
          >
            <rect
              height="13"
              width="15"
              x="1"
              y="3"
            />
            <polygon
              points="16 8 20 8 23 11 23 16 16 16 16 8"
            />
            <circle
              cx="5.5"
              cy="18.5"
              r="2.5"
            />
            <circle
              cx="18.5"
              cy="18.5"
              r="2.5"
            />
          </svg>
        </div>
      </div>
      <h2
        class="c6 ProductCard__title"
      >
        Mocked product title
      </h2>
    </div>
    <div
      class="c7"
    >
      <span
        class="c8"
      >
        Produto novo
      </span>
    </div>
  </a>
</div>
`;
